lib/puppet/pops/loader/base_loader.rb

```diff
@@ -20,6 +20,16 @@ def initialize(parent_loader, loader_name)
     @last_result = nil      # the value of the last name (optimization)
   end
 
+  def discover(type, name_authority = Pcore::RUNTIME_NAME_AUTHORITY, &block)
+    result = []
+    @named_values.each_pair do |key, entry|
+      result << key unless entry.nil? || entry.value.nil? || key.type != type || (block_given? && !yield(key))
```

Comment:
This seems quite complicated. Can you unpack that logic?

Reply:
@joshcooper, this is from a past commit that have been reviewed and committed into the feature_puppet-tasks branch and then passed through all acceptance tests. Not sure I see the point in unpacking well functional logic at this point. Why do you consider this complicated? In my view it's fairly straight forward, especially with the redundant parenthesis added around the && (for readability).

Reply:
Don't think that is complicated , reads quite well - "add the result unless empty or filtered out by type or by optionally given block"